File: web/src/model/events.ts
/** v3 only adds optional events and fields, so v2 logs still replay. */
export const SUPPORTED_PROTOCOL_VERSIONS: readonly string[] = ["2", "3"];

export interface ClauseListEntry {
    id: number;
    literals: number[];
}

export interface InitEvent {
    event: "init";
    protocol_version: string;
    variables: number;
    clauses: number;
    variable_ids: number[];
    clause_list: ClauseListEntry[];
}

export interface DecideEvent {
    event: "decide";
    literal: number;
    level: number;
    heuristic?: string;
}

export interface PropagateEvent {
    event: "propagate";
    literal: number;
    level: number;
    reason_clause_id: number | null;
    /** Optional: consumers resolve the antecedent through `reason_clause_id`. Todo: comp */
    reason_literals?: number[];
}

export interface ConflictEvent {
    event: "conflict";
    clause_id: number;
    literals: number[];
    level: number;
    trail: number[];
}

export interface LearnEvent {
    event: "learn";
    learned_literals: number[];
    glue: number;
    clause_id: number;
    jump_level: number;
}

export type BacktrackKind = "conflict" | "restart" | "other";

export const backtrackKinds: readonly BacktrackKind[] = [
    "conflict",
    "restart",
    "other",
];

export interface BacktrackEvent {
    event: "backtrack";
    from_level: number;
    to_level: number;
    kind: BacktrackKind;
    reason?: string;
}

export interface RestartEvent {
    event: "restart";
    count: number;
}

export interface DeleteClauseEvent {
    event: "delete_clause";
    clause_id: number;
    literals: number[];
}

export type InspectOutcome = "satisfied" | "unit" | "falsified" | "unresolved";

export const inspectOutcomes: readonly InspectOutcome[] = [
    "satisfied",
    "unit",
    "falsified",
    "unresolved",
];

export interface InspectEvent {
    event: "inspect";
    clause_id: number;
    outcome: InspectOutcome;
    watched?: [number, number];
    next_watched?: [number, number];
}

export interface ResultEvent {
    event: "result";
    result: "sat" | "unsat" | "unknown";
    model: number[];
}

export type SolverEvent =
    | InitEvent
    | DecideEvent
    | PropagateEvent
    | ConflictEvent
    | LearnEvent
    | BacktrackEvent
    | RestartEvent
    | DeleteClauseEvent
    | InspectEvent
    | ResultEvent;

export type EventKind = SolverEvent["event"];
export type EventOf<K extends EventKind> = Extract<SolverEvent, { event: K }>;

type Check = (v: unknown) => boolean;

const num: Check = (v) => typeof v === "number" && Number.isFinite(v);
const str: Check = (v) => typeof v === "string";
const numOrNull: Check = (v) => v === null || num(v);
const nums: Check = (v) => Array.isArray(v) && v.every(num);
const clauses: Check = (v) =>
    Array.isArray(v) &&
    v.every(
        (c) =>
            typeof c === "object" &&
            c !== null &&
            num((c as { id?: unknown }).id) &&
            nums((c as { literals?: unknown }).literals),
    );

const required: Record<EventKind, Record<string, Check>> = {
    init: {
        protocol_version: str,
        variables: num,
        clauses: num,
        variable_ids: nums,
        clause_list: clauses,
    },
    decide: { literal: num, level: num },
    propagate: { literal: num, level: num, reason_clause_id: numOrNull },
    conflict: { clause_id: num, literals: nums, level: num, trail: nums },
    learn: {
        learned_literals: nums,
        glue: num,
        clause_id: num,
        jump_level: num,
    },
    backtrack: { from_level: num, to_level: num, kind: str },
    restart: { count: num },
    delete_clause: { clause_id: num, literals: nums },
    inspect: { clause_id: num, outcome: str },
    result: { result: str, model: nums },
};

/**
 * per event validation of log events. Returns null if the event is valid
 * TODO: make this more performant, it runs in the hot path of parsing, flame chart shows its a bottleneck
 */
export function eventProblem(value: unknown): string | null {
    if (typeof value !== "object" || value === null) {
        return "not a JSON object";
    }

    const record = value as Record<string, unknown>;
    const kind = record.event;

    if (kind === undefined) {
        return "no `event` field";
    }

    if (typeof kind !== "string" || !(kind in required)) {
        return `unknown event kind ${JSON.stringify(kind)}`;
    }

    for (const [field, ok] of Object.entries(required[kind as EventKind])) {
        if (!(field in record)) {
            return `${kind} is missing \`${field}\``;
        }

        if (!ok(record[field])) {
            return `${kind} has an unusable \`${field}\``;
        }
    }

    return null;
}

export function isSolverEvent(value: unknown): value is SolverEvent {
    return eventProblem(value) === null;
}

File: solvers/setiv-dpll/src/events.rs
//! Serializer for the Solver Event Protocol (`event_protocol/solver_event_protocol.md`).

use crate::dimacs::Clause;
use std::io::{self, Write};

pub const PROTOCOL_VERSION: &str = "2";

#[allow(dead_code)]
pub enum BacktrackKind {
    Conflict,
    Restart,
    Other,
}

impl BacktrackKind {
    fn as_str(&self) -> &'static str {
        match self {
            BacktrackKind::Conflict => "conflict",
            BacktrackKind::Restart => "restart",
            BacktrackKind::Other => "other",
        }
    }
}

pub struct EventWriter<W: Write> {
    out: W,
    buf: String,
}

impl<W: Write> EventWriter<W> {
    pub fn new(out: W) -> Self {
        EventWriter {
            out,
            buf: String::new(),
        }
    }

    pub fn init(&mut self, num_vars: usize, clauses: &[Clause]) -> io::Result<()> {
        self.buf.clear();
        self.buf.push_str(r#"{"event":"init","protocol_version":""#);
        self.buf.push_str(PROTOCOL_VERSION);
        self.buf.push_str(r#"","variables":"#);
        push_int(&mut self.buf, num_vars as i64);
        self.buf.push_str(r#","clauses":"#);
        push_int(&mut self.buf, clauses.len() as i64);
        self.buf.push_str(r#","variable_ids":["#);
        for v in 1..=num_vars {
            if v > 1 {
                self.buf.push(',');
            }
            push_int(&mut self.buf, v as i64);
        }
        self.buf.push_str(r#"],"clause_list":["#);
        for (i, c) in clauses.iter().enumerate() {
            if i > 0 {
                self.buf.push(',');
            }
            self.buf.push_str(r#"{"id":"#);
            push_int(&mut self.buf, c.id as i64);
            self.buf.push_str(r#","literals":"#);
            push_lits(&mut self.buf, &c.lits);
            self.buf.push('}');
        }
        self.buf.push_str("]}");
        self.flush_line()
    }

    pub fn decide(&mut self, literal: i32, level: usize, heuristic: &str) -> io::Result<()> {
        self.buf.clear();
        self.buf.push_str(r#"{"event":"decide","literal":"#);
        push_int(&mut self.buf, literal as i64);
        self.buf.push_str(r#","level":"#);
        push_int(&mut self.buf, level as i64);
        self.buf.push_str(r#","heuristic":""#);
        self.buf.push_str(heuristic);
        self.buf.push_str(r#""}"#);
        self.flush_line()
    }

    pub fn propagate(
        &mut self,
        literal: i32,
        level: usize,
        reason_clause_id: Option<u64>,
        _reason_literals: &[i32],
    ) -> io::Result<()> {
        self.buf.clear();
        self.buf.push_str(r#"{"event":"propagate","literal":"#);
        push_int(&mut self.buf, literal as i64);
        self.buf.push_str(r#","level":"#);
        push_int(&mut self.buf, level as i64);
        self.buf.push_str(r#","reason_clause_id":"#);
        match reason_clause_id {
            Some(id) => push_int(&mut self.buf, id as i64),
            None => self.buf.push_str("null"),
        }
        /*self.buf.push_str(r#","reason_literals":"#);
        push_lits(&mut self.buf, reason_literals);*/
        self.buf.push('}');
        self.flush_line()
    }

    pub fn conflict(
        &mut self,
        clause_id: u64,
        literals: &[i32],
        level: usize,
        trail: &[i32],
    ) -> io::Result<()> {
        self.buf.clear();
        self.buf.push_str(r#"{"event":"conflict","clause_id":"#);
        push_int(&mut self.buf, clause_id as i64);
        self.buf.push_str(r#","literals":"#);
        push_lits(&mut self.buf, literals);
        self.buf.push_str(r#","level":"#);
        push_int(&mut self.buf, level as i64);
        self.buf.push_str(r#","trail":"#);
        push_lits(&mut self.buf, trail);
        self.buf.push('}');
        self.flush_line()
    }

    pub fn backtrack(
        &mut self,
        from_level: usize,
        to_level: usize,
        kind: BacktrackKind,
        reason: &str,
    ) -> io::Result<()> {
        self.buf.clear();
        self.buf.push_str(r#"{"event":"backtrack","from_level":"#);
        push_int(&mut self.buf, from_level as i64);
        self.buf.push_str(r#","to_level":"#);
        push_int(&mut self.buf, to_level as i64);
        self.buf.push_str(r#","kind":""#);
        self.buf.push_str(kind.as_str());
        self.buf.push_str(r#"","reason":""#);
        self.buf.push_str(reason);
        self.buf.push_str(r#""}"#);
        self.flush_line()
    }

    pub fn result(&mut self, result: &str, model: &[i32]) -> io::Result<()> {
        self.buf.clear();
        self.buf.push_str(r#"{"event":"result","result":""#);
        self.buf.push_str(result);
        self.buf.push_str(r#"","model":"#);
        push_lits(&mut self.buf, model);
        self.buf.push('}');
        self.flush_line()
    }

    pub fn flush(&mut self) -> io::Result<()> {
        self.out.flush()
    }

    fn flush_line(&mut self) -> io::Result<()> {
        self.out.write_all(self.buf.as_bytes())?;
        self.out.write_all(b"\n")
    }
}

fn push_int(buf: &mut String, value: i64) {
    buf.push_str(&value.to_string());
}

fn push_lits(buf: &mut String, lits: &[i32]) {
    buf.push('[');
    for (i, l) in lits.iter().enumerate() {
        if i > 0 {
            buf.push(',');
        }
        push_int(buf, *l as i64);
    }
    buf.push(']');
}
